### PR TITLE
[interop][SwiftToCxx] Pass struct to function as inout

### PR DESCRIPTION
**Pull request by WANGJIEKE**
This pull request supports passing a `struct` type into a function as an `inout` parameter.

@hyp Could you please review this pull request?

Here is an overview of changes I've made:

- When the parameter is `inout`, the Swift struct `MyStruct` will be printed as `MyStruct &` in the C++ function signature (the `const` qualifier is removed compared to a non-`inout` parameter).

- If the struct is a "small struct" (i.e., need direct pass):

  - In `inout` mode, the underlying C interface will take a `char * _Nonnull` directly instead of a C stub struct value (similar to `inout` mode for primitives).

  - In `inout` mode, the C++ function body will pass the pointer obtained from `getOpaquePointer()` directly to the underlying C interface instead of making a C stub struct first and passing that stub by value.

  - Please point out if I did something wrong here. This approach passed tests, but I am not sure if I can pass the pointer directly, or I need to make that C stub struct a local variable, pass the address of the C stub struct, and somehow copy the result back to the C++ struct.

- If the struct is a "large struct" (i.e., need indirect pass):

  - In `inout` mode, the underlying C interface will take a `void * _Nonnull` instead of `const void * _Nonnull` in non-`inout` mode.

- Add some parameters to `CFunctionSignatureTypePrinter` class, `ClangValueTypePrinter::printValueTypeParameterType` and `ClangValueTypePrinter::printParameterCxxToCUseScaffold` functions to support the new feature

- Add some tests

### CODE REVIEW

File: lib/PrintAsClang/PrintClangFunction.cpp [hunk 1]
```diff
@@ -157,10 +166,16 @@ void DeclAndTypeClangFunctionPrinter::printFunctionSignature(
     // see DeclAndTypePrinter::print.
     CFunctionSignatureTypePrinter typePrinter(os, cPrologueOS, typeMapping,
                                               outputLang, interopContext);
-    typePrinter.visit(ty, optionalKind);
+    typePrinter.visit(ty, optionalKind, isInOutParam);
 
     if (isInOutParam) {
-      os << (outputLang == OutputLanguageMode::Cxx ? " &" : " *");
+      if (outputLang == OutputLanguageMode::Cxx &&
+          isKnownCxxType(ty, typeMapping)) {
```
Comment (hyp) [inline lib/PrintAsClang/PrintClangFunction.cpp:173]: Suggestion: can you move the `&` / `*` printer for primitive types into the visitor as well, to be more consistent with the inout types for struct types?  You can add the `bool` flag to `printIfKnownSimpleType` and handle it in there.

File: lib/PrintAsClang/PrintClangFunction.cpp [hunk 2]
```diff
@@ -182,7 +197,9 @@ void DeclAndTypeClangFunctionPrinter::printFunctionSignature(
     CFunctionSignatureTypePrinter typePrinter(
         os, cPrologueOS, typeMapping, outputLang, interopContext,
         FunctionSignatureTypeUse::ReturnType);
-    typePrinter.visit(objTy, retKind);
+    // Pass false because it's a parameter for indirect return.
+    // Not an actual parameter, which means it can never be inout
+    typePrinter.visit(objTy, retKind, false);
```
Comment (hyp) [inline lib/PrintAsClang/PrintClangFunction.cpp:202]: NIT: You can add a comment before `false` to explain the param - `/*isInOutParam=*/false`, and then you can shorten the comments you have above it to reduce the justification.

**Review by hyp — COMMENTED** [review]
Overall looks great, just some minor requests!

**Comment by WANGJIEKE** [comment]
Updated :)

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

File: lib/PrintAsClang/PrintClangFunction.cpp [hunk 3]
```diff
@@ -77,62 +77,75 @@ class CFunctionSignatureTypePrinter
...
+                              Optional<OptionalTypeKind> optionalKind,
+                              bool isInOutParam) {
     auto knownTypeInfo = getKnownTypeInfo(typeDecl, typeMapping, languageMode);
     if (!knownTypeInfo)
       return false;
     os << knownTypeInfo->name;
-    if (knownTypeInfo->canBeNullable)
+    if (knownTypeInfo->canBeNullable) {
       printNullability(optionalKind);
+    } else if (isInOutParam) {
```
Comment (hyp) [inline lib/PrintAsClang/PrintClangFunction.cpp:88]: I don't think this should be an `else if`, but rather an `if`. You can have a nullable `inout`, e.g. `inout UnsafeRawPointer`. Can you fix it and add a test to catch it?

**Review by hyp — COMMENTED** [review]
One more request.

File: lib/PrintAsClang/PrintClangFunction.cpp (hunk 3 above)
Comment (WANGJIEKE) [reply lib/PrintAsClang/PrintClangFunction.cpp:88]: Sure

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please smoke test macos

**Comment by hyp** [comment]
Failed in `Swift(macosx-x86_64) :: SIL/Parser/moveonly_to_copyable.sil` on macOS, this is unrelated. MacOS smoke test passes, so I will merge.

**Pull request merged**